File: jiyeong/week_01/PROG_42579_베스트앨범.py
def solution(genres, plays):
    answer = []
    gMap = dict()
    N = len(plays)

    # 각 장르별로 노래 구분하기 => {장르 : [(재생횟수, 고유번호)]}
    for i in range(N):
        if genres[i] not in gMap.keys():
            gMap[genres[i]] = [(plays[i], i)]
        else:
            gMap[genres[i]].append((plays[i], i))
    # print(gMap)

    # 장르별 총 재생횟수 구하기 => [(총재생횟수, 장르명)]
    gList = []
    for k in gMap.keys():
        pSum = 0
        for p in gMap[k]:
            pSum += p[0]
        gList.append((pSum, k))
    gList.sort(key=lambda x: x[0], reverse=True)

    for i in range(len(gList)):
        pList = gMap[gList[i][1]]
        # print(pList)
        if len(pList) == 1:
            answer.append(pList[0][1])
            continue
        pList.sort(key=lambda x: x[0], reverse=True)
        answer.append(pList[0][1])
        answer.append(pList[1][1])

    return answer

File: hanbin/week_01/PROG_42579_베스트앨범.js
// 풀이 전략: 해싱, 정렬 O(n^2)

function solution(genres, plays) {
    var answer = [];
    
    let map = new Map();
    
    // 장르별로 map에 저장
    genres.forEach((genre, idx) => {
        const play = plays[idx];
        
        if (!map.has(genre)){
            map.set(genre, []);
        }
        
        map.get(genre).push({num: idx, play});
    })
    
    // 장르 내에서 재생 횟수 순으로 정렬
    for (let key of map.keys()){
        let arr = map.get(key);
        arr.sort((a, b) => {
            if (a.play !== b.play){
                return b.play - a.play;
            } else {
                return a.num - b.num;
            }
        })
    }
    
    // 장르끼리 비교해서 재생횟수 순으로 정렬
    map = new Map([...map].sort((a, b) => {
        const sumA = a[1].reduce((prev, cur) => prev + cur.play, 0);
        const sumB = b[1].reduce((prev, cur) => prev + cur.play, 0);
        
        return sumB - sumA;
    }));
    
    // 정렬된 장르 순회하면서 탑2까지만 answer 배열에 저장
    for (let genre of map){        
        genre[1].forEach((it, idx) => {
            if (idx >= 2) return;
            answer.push(it.num);
        })
    }
    
    return answer;
}
